### PR TITLE
Conditions model persistence

### PR DESCRIPTION
**Pull request by andymoody**
# Description

Storing conditions as a structured data model.

- Make conditions model serializable and deserializable as json
- Adapt the schema to handle both string and structured data versions of conditions
- Break out a separate 'model' library to store shared items
- move component-types, conditional-component-types, and schema into the shared model.
- move the conditions model to the new model library
- Use the conditions model in engine to support evaluation of structured data conditions

## Type of change

Please delete options that are not relevant.

- [X] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce.

- [X] Unit tests
- [X] Click testing

# Checklist:

- [X] My code follows the [javascript standard style](https://standardjs.com/)
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation and versioning
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] I have rebased onto master and there are no code conflicts

### CODE REVIEW

File: designer/client/component-condition-create.js [hunk 1]
```diff
@@ -1,6 +1,6 @@
 import React from 'react'
 import ComponentTypeEdit from './component-type-edit'
-import ConditionalComponentTypes from 'digital-form-builder-engine/src/conditional-component-types'
+import ConditionalComponentTypes from 'digital-form-builder-model/lib/conditional-component-types'
```
Comment (jenbutongit) [inline designer/client/component-condition-create.js:3]: sorry I didn't catch this before..! take a look at `-model/src/lib`. I'm exporting everything that's meant to be consumed in the package. Then in in `model/index`, I'm importing (the transpiled) lib.js. This way you don't have to do (or shouldn't?) import `'digital-form-builder-model/lib`. you can just do `import ... from digital-form-builder-model`. This is more in line with how npm packages are consumed

File: designer/client/conditions/inline-conditions-definition-values.js [hunk 2]
```diff
@@ -1,6 +1,12 @@
 import React from 'react'
-import { ConditionValue } from './inline-condition-values'
-import { getOperatorConfig } from './inline-condition-operators'
+import { ConditionValue, timeUnits } from 'digital-form-builder-model/lib/conditions/inline-condition-values'
+import {
+  absoluteDateOrTimeOperatorNames,
+  getOperatorConfig, relativeDateOrTimeOperatorNames
+} from 'digital-form-builder-model/lib/conditions/inline-condition-operators'
+
```
Comment (jenbutongit) [inline designer/client/conditions/inline-conditions-definition-values.js:7]: see comment for `component-condition-create.js`

File: designer/client/conditions/inline-conditions-definition.js [hunk 3]
```diff
@@ -1,9 +1,9 @@
 import React from 'react'
-import { Condition, Field } from './inline-condition-model'
-import { valueFrom } from './inline-condition-values'
-import { getOperatorNames } from './inline-condition-operators'
-import { clone } from '../helpers'
+import { Condition, Field } from 'digital-form-builder-model/lib/conditions/inline-condition-model'
+import { valueFrom } from 'digital-form-builder-model/lib/conditions/inline-condition-values'
+import { getOperatorNames } from 'digital-form-builder-model/lib/conditions/inline-condition-operators'
```
Comment (jenbutongit) [inline designer/client/conditions/inline-conditions-definition.js:4]: see comment for `component-condition-create.js`

File: designer/client/conditions/inline-conditions-edit.js [hunk 4]
```diff
@@ -1,8 +1,8 @@
 import React from 'react'
-import { GroupDef } from './inline-condition-model'
-import { clone } from '../helpers'
+import { GroupDef } from 'digital-form-builder-model/lib/conditions/inline-condition-model'
 import { icons } from '../icons'
 import InlineConditionsDefinition from './inline-conditions-definition'
+import { clone } from 'digital-form-builder-model/lib/helpers'
```
Comment (jenbutongit) [inline designer/client/conditions/inline-conditions-edit.js:5]: see comment for `component-condition-create.js`

File: designer/client/conditions/inline-conditions-relative-dates.js [hunk 5]
```diff
@@ -1,25 +1,5 @@
...
-      if (value instanceof RelativeTimeValue) {
-        const operator = value.direction === dateDirections.PAST ? pastOperator : futureOperator
-        return `${field.name} ${operator} ${value.toExpression()}`
-      }
-      throw Error('time shift requires a TimeShiftValue')
-    },
-    renderComponent: (value, updateValue) => <RelativeTimeValues value={value} updateValue={updateValue} units={units} timeOnly={units === timeUnits} />
-  }
-}
+import { dateDirections, RelativeTimeValue } from 'digital-form-builder-model/lib/conditions/inline-condition-values'
```
Comment (jenbutongit) [inline designer/client/conditions/inline-conditions-relative-dates.js:2]: see comment for `component-condition-create.js`

File: designer/client/conditions/inline-conditions.js [hunk 6]
```diff
@@ -1,8 +1,8 @@
 import React from 'react'
-import { clone } from '../helpers'
-import { ConditionsModel } from './inline-condition-model'
+import { ConditionsModel } from 'digital-form-builder-model/lib/conditions/inline-condition-model'
 import InlineConditionsDefinition from './inline-conditions-definition'
 import InlineConditionsEdit from './inline-conditions-edit'
+import { clone } from 'digital-form-builder-model/lib/helpers'
```
Comment (jenbutongit) [inline designer/client/conditions/inline-conditions.js:5]: see comment for `component-condition-create.js`

File: designer/client/conditions/select-conditions.js [hunk 7]
```diff
@@ -1,6 +1,6 @@
 import React from 'react'
 import InlineConditions from './inline-conditions'
-import { ConditionsModel } from './inline-condition-model'
+import { ConditionsModel } from 'digital-form-builder-model/lib/conditions/inline-condition-model'
```
Comment (jenbutongit) [inline designer/client/conditions/select-conditions.js:3]: see comment for `component-condition-create.js`

File: designer/client/fee-items.js [hunk 8]
```diff
@@ -1,5 +1,5 @@
 import React from 'react'
-import { clone } from './helpers'
+import { clone } from 'digital-form-builder-model/lib/helpers'
```
Comment (jenbutongit) [inline designer/client/fee-items.js:2]: i'll stop mentioning `see comment for component-condition-create.js` but you get the idea..!

File: model/test/data-model.test.js [hunk 9]
```diff
@@ -1,7 +1,7 @@
 import * as Code from '@hapi/code'
 import * as Lab from '@hapi/lab'
 
-import { Data } from '../client/model/data-model'
+import { Data } from '../src/data-model'
```
Comment (jenbutongit) [inline model/test/data-model.test.js:4]: should probably be using exported { Data } from (root) index.js?

File: model/readme.md [hunk 10]
```diff
@@ -0,0 +1,21 @@
+# digital-form-builder-model
+
+A hapi plugin providing the engine for [digital form builder](https://github.com/UKForeignOffice/digital-form-builder) based applications. 
+
+To use this plugin it must be registered using the server in UKForeignOffice/digital-form-builder.
```
[the remark this answers is not in the document]
Comment (andymoody) [reply model/readme.md:5]: Done

File: model/src/data-model.js [hunk 11]
```diff
@@ -205,4 +206,20 @@ class Condition {
     Object.assign(this, rawData)
     this.displayName = rawData.displayName || rawData.name
   }
+
+  get expression () {
+    if (typeof this.value === 'string') {
+      return this.value
+    } else {
+      const conditions = ConditionsModel.from(this.value)
+      return conditions.toExpression()
```
[the remark this answers is not in the document]
Comment (andymoody) [reply model/src/data-model.js:215]: Done

File: model/.babelrc [hunk 12]
```diff
@@ -0,0 +1,11 @@
+{
+  "presets": ["@babel/preset-env"],
```
[the remark this answers is not in the document]
Comment (andymoody) [reply model/.babelrc:2]: Done

File: designer/client/component-condition-create.js (hunk 1 above)
Comment (andymoody) [reply designer/client/component-condition-create.js:3]: Fixed

File: model/test/data-model.test.js (hunk 9 above)
Comment (andymoody) [reply model/test/data-model.test.js:4]: Fixed

File: designer/client/conditions/inline-condition-helpers.js [hunk 13]
```diff
@@ -1,8 +1,8 @@
-export async function storeConditionIfNecessary (data, conditions) {
+async function storeConditionIfNecessary (data, conditions) {
   let condition
   if (conditions && conditions.hasConditions) {
     condition = await data.getId()
```
Comment (jenbutongit) [inline designer/client/conditions/inline-condition-helpers.js:4]: I must've missed this. feel free to resolve in another PR. data.getId() looks like it's getting the id of the data object. maybe this should be renamed generateId?

File: designer/client/conditions/inline-conditions-definition-values.js [hunk 14]
```diff
@@ -49,12 +55,31 @@ function SelectValues (props) {
   )
 }
 
+function customValueComponent (fieldType, operator) {
+  const operatorConfig = getOperatorConfig(fieldType, operator)
+  const absoluteDateTimeRenderFunctions = {
+    'DateField': (value, updateValue) => <AbsoluteDateValues value={value} updateValue={updateValue} />,
```
Comment (jenbutongit) [inline designer/client/conditions/inline-conditions-definition-values.js:61]: do the keys need to be strings?

File: designer/client/conditions/inline-conditions-definition-values.js [hunk 15]
```diff
@@ -49,12 +55,31 @@ function SelectValues (props) {
...
+    'DatePartsField': (value, updateValue) => <AbsoluteDateValues value={value} updateValue={updateValue} />,
+    'DateTimeField': (value, updateValue) => <AbsoluteDateTimeValues value={value} updateValue={updateValue} />,
+    'DateTimePartsField': (value, updateValue) => <AbsoluteDateTimeValues value={value} updateValue={updateValue} />,
+    'TimeField': (value, updateValue) => <AbsoluteTimeValues value={value} updateValue={updateValue} />
+  }
+  const dateTimeFieldTypes = Object.keys(absoluteDateTimeRenderFunctions).includes(fieldType)
```
Comment (jenbutongit) [inline designer/client/conditions/inline-conditions-definition-values.js:67]: `absoluteDateTimeRenderFunctions[fieldType]`? 

`includes` will only return a bool, the const name makes it sound like an array of fieldTypes

File: designer/client/conditions/inline-conditions-definition-values.js [hunk 16]
```diff
@@ -49,12 +55,31 @@ function SelectValues (props) {
   )
 }
 
+function customValueComponent (fieldType, operator) {
```
Comment (jenbutongit) [inline designer/client/conditions/inline-conditions-definition-values.js:58]: I think you're trying to do this:
https://reactjs.org/docs/jsx-in-depth.html#choosing-the-type-at-runtime

this fn returning components should really be it's own component

File: model/readme.md [hunk 17]
```diff
@@ -0,0 +1,21 @@
+# digital-form-builder-model
+
+A hapi plugin providing the engine for [digital form builder](https://github.com/XGovFormBuilder/digital-form-builder) based applications. 
+
+To use this plugin it must be registered using the server in XGovFormBuilder/digital-form-builder.
+
```
Comment (jenbutongit) [inline model/readme.md:6]: is this still correct?

File: model/src/conditions/inline-condition-model.js [hunk 18]
```diff
@@ -194,6 +194,29 @@ export class ConditionsModel {
     }
     return []
   }
+
+  toJSON () {
```
Comment (jenbutongit) [inline model/src/conditions/inline-condition-model.js:198]: this isn't a returning JSON. do you mean to toObject? or to wrap in the return in `JSON.stringify(..)`?

File: model/src/conditions/inline-condition-operators.js [hunk 19]
```diff
@@ -92,3 +103,30 @@ function formatValue (fieldType, value) {
   }
   return `'${value}'`
```
Comment (jenbutongit) [inline model/src/conditions/inline-condition-operators.js:104]: must've missed this in another PR, why are you returning string within a string?

File: model/src/conditions/inline-condition-operators.js [hunk 20]
```diff
@@ -92,3 +103,30 @@ function formatValue (fieldType, value) {
...
 }
+
+export const absoluteDateOrTimeOperatorNames = Object.keys(absoluteDateTimeOperators)
+export const relativeDateOrTimeOperatorNames = Object.keys(relativeTimeOperators(dateTimeUnits))
+
+function absoluteDateTime (operator) {
+  return {
+    expression: (field, value) => {
+      if (value instanceof ConditionValue) {
+        return `${field.name} ${operator} '${value.toExpression()}'`
```
Comment (jenbutongit) [inline model/src/conditions/inline-condition-operators.js:114]: should toExpression be a getter? `value.expression`

File: designer/client/conditions/inline-condition-helpers.js (hunk 13 above)
Comment (andymoody) [reply designer/client/conditions/inline-condition-helpers.js:4]: I'll resolve this another time, but yes, good point!

File: designer/client/conditions/inline-conditions-definition-values.js (hunk 14 above)
Comment (andymoody) [reply designer/client/conditions/inline-conditions-definition-values.js:61]: Fixed

File: designer/client/conditions/inline-conditions-definition-values.js (hunk 15 above)
Comment (andymoody) [reply designer/client/conditions/inline-conditions-definition-values.js:67]: Fixed

File: designer/client/conditions/inline-conditions-definition-values.js (hunk 16 above)
Comment (andymoody) [reply designer/client/conditions/inline-conditions-definition-values.js:58]: Sort of, but not - as the props also differ between the various things it returns. I've amended to something closer to this though.

File: model/readme.md (hunk 17 above)
Comment (andymoody) [reply model/readme.md:6]: Fixed

File: model/src/conditions/inline-condition-model.js (hunk 18 above)
Comment (andymoody) [reply model/src/conditions/inline-condition-model.js:198]: No, this is part of the JSON.stringify functionality. In order to customise the serialisation you define a toJSON method returning the object you want to be serialised: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#toJSON_behavior

File: model/src/conditions/inline-condition-operators.js (hunk 19 above)
Comment (andymoody) [reply model/src/conditions/inline-condition-operators.js:104]: The resultant expr-eval expression needs to look something like "field = 'value'" - i.e. string values need to be wrapped in something identifying it as such.

File: model/src/conditions/inline-condition-operators.js (hunk 20 above)
Comment (andymoody) [reply model/src/conditions/inline-condition-operators.js:114]: Possibly, this feels like perhaps a nice to have that we can address in a later pr?

File: designer/client/conditions/inline-conditions-definition-values.js (hunk 16 above)
Comment (jenbutongit) [reply designer/client/conditions/inline-conditions-definition-values.js:58]: the props you're passing into the components are derived from fieldType and operator, and at least absoluteDateTimeRenderFunctions all accept the same properties (or prop names). All `InlineConditionsDefinitionValue` knows is that c(C)ustomValueComponent is returning a component

File: model/src/conditions/inline-condition-model.js (hunk 18 above)
Comment (jenbutongit) [reply model/src/conditions/inline-condition-model.js:198]: oh, that's interesting!

**Comment by jenbutongit** [comment]
squash merge please!

**Pull request merged**